### PR TITLE
implement http3.RoundTripper.CloseIdleConnections

### PR DESCRIPTION
**Pull request by Glonee**
http: implement http3.RoundTripper.CloseIdleConnections #3814

### CODE REVIEW

**Comment by Glonee** [comment]
I am not sure my implemention is correct. In the standard library, the `http2ClientConn` has a method called `closeIfIdle`, see https://github.com/golang/go/blob/a25688d406f2a4296d39b22a2b10aea7178eddc6/src/net/http/h2_bundle.go#L8044-L8059.
I didn't find any similar method in quic.Connection nor method to find out if a connection is idle. So I count the use of every connection, and those have zero count are considered to be idle.
Let me know if you have any thought about this.

File: http3/roundtrip.go [hunk 1]
```diff
@@ -204,10 +211,13 @@ func (r *RoundTripper) getClient(hostname string, onlyCached bool) (rtc roundTri
 		if err != nil {
 			return nil, false, err
 		}
+		client = new(roundTripCloserWithCount)
+		client.roundTripCloser = c
```
Comment (marten-seemann) [inline http3/roundtrip.go:215]: Nit:
```suggestion
		client = &roundTripCloserWithCount{roundTripCloser: c}
```

File: http3/roundtrip.go [hunk 2]
```diff
@@ -25,6 +26,11 @@ type roundTripCloser interface {
 	io.Closer
 }
 
+type roundTripCloserWithCount struct {
+	roundTripCloser
+	useCount atomic.Int64
```
Comment (marten-seemann) [inline http3/roundtrip.go:31]: Both `getClient` and `CloseIdleConnections` run with the `Roundtripper.mutex` locked. Would it be sufficient to make this an `int`?

**Review by marten-seemann — COMMENTED** [review]
Thank you @Glonee! Just a few comments, this mostly looks good.

Counting the number of active `RoundTrip` calls seems correct. The server can't initiate any streams in HTTP/3 (other than server push, which we don't support).

File: http3/roundtrip.go [hunk 3]
```diff
@@ -143,6 +149,7 @@ func (r *RoundTripper) RoundTripOpt(req *http.Request, opt RoundTripOpt) (*http.
 	if err != nil {
 		return nil, err
 	}
+	defer cl.useCount.Add(-1)
```
Comment (Glonee) [inline http3/roundtrip.go:152]: We reduce the count here, out of the lock. Unless we add a lock here, or we should use atomic.

File: http3/roundtrip.go (hunk 2 above)
Comment (Glonee) [reply http3/roundtrip.go:31]: https://github.com/quic-go/quic-go/blob/b6c36841a5e5063864b628dd2beab780a8d043f2/http3/roundtrip.go#L152
We reduce the count here, out of the lock. Unless we add a lock here, or we should use atomic.

Comment (marten-seemann) [reply http3/roundtrip.go:31]: Makes sense, thank you!

**Review by marten-seemann — APPROVED** [review]
LGTM, thank you @Glonee!

**Pull request merged**